<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''SELECT COUNT(*)
                FROM vacancies
             '''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''SELECT COUNT(*)
                FROM employers
            '''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''SELECT COUNT(*)
                FROM areas
            '''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''SELECT COUNT(*)
                FROM industries
             '''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    294

***

### Выводы по предварительному анализу данных:
В данных представлены 49197 вакансий, работодатели, которых почти в 2 раза меньше вакансий(23501), 1362 различных региона и 294 сфер деятельности. 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = '''SELECT COUNT(*) cnt, a.name area
               FROM vacancies v JOIN areas a ON v.area_id=a.id
               GROUP BY area
               ORDER BY 1 DESC'''

In [13]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt             area
0    5333           Москва
1    2851  Санкт-Петербург
2    2112            Минск
3    2006      Новосибирск
4    1892           Алматы
..    ...              ...
764     1       Тарко-Сале
765     1    Новоаннинский
766     1            Бирск
767     1           Сасово
768     1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = '''SELECT COUNT(*)
               FROM vacancies
               WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL'''

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = '''SELECT AVG(salary_from)::integer salary_from, AVG(salary_to)::integer salary_to 
               FROM vacancies -- приводим к integer, чтобы округлить значения до целого.'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


salary_from  salary_to
0        71065     110537

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = '''SELECT schedule, employment, COUNT(*) cnt_vacancies
               FROM vacancies 
               GROUP BY schedule, employment
               ORDER BY cnt_vacancies DESC'''

In [19]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule           employment  cnt_vacancies
0        Полный день     Полная занятость          35367
1   Удаленная работа     Полная занятость           7802
2      Гибкий график     Полная занятость           1593
3   Удаленная работа  Частичная занятость           1312
4     Сменный график     Полная занятость            940
5        Полный день           Стажировка            569
6     Вахтовый метод     Полная занятость            367
7        Полный день  Частичная занятость            347
8      Гибкий график  Частичная занятость            312
9        Полный день     Проектная работа            141
10  Удаленная работа     Проектная работа            133
11     Гибкий график           Стажировка            116
12    Сменный график  Частичная занятость            101
13  Удаленная работа           Стажировка             64
14     Гибкий график     Проектная работа             18
15    Сменный график           Стажировка             12
16    Вахтовый метод     Проектная работа              2
17    Сменный график     Проектная работа              1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = '''SELECT experience, COUNT(*) cnt_vacancies
               FROM vacancies
               GROUP BY experience
               ORDER BY cnt_vacancies '''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  cnt_vacancies
0         Более 6 лет           1337
1           Нет опыта           7197
2       От 3 до 6 лет          14511
3  От 1 года до 3 лет          26152

***

### Выводы по детальному анализу вакансий:
1. Ожидаемо большее количество вакансий в крупных городах по типу Москвы, Санкт-Петербурга и Минска.
2. Больше всего вакансий, где указан опыт работы от 1 года до 3 лет, что можно объяснить тем, что работодателем выгоднее всего взять не совсем опытного работника, чтобы не платить ему много.
3. Вакансий, где не нужен опыт работы, не так много, т.к. мало кому хочется возиться с людьми, которые плохо понимают, что от них просят.
4. Меньше всего вакансий, где требуется опыт работы более 6 лет. Скорее всего, это связано с тем, что такому ценному сотруднику придётся платить большую зарплату, поэтому работадатели пытаются искать таких людей только в крайних случаях. Больше всего вакансий, где требуется полная занятость и полный день.
5. Всего 24073(это меньше половины) вакансий, где указана зарплата. Было бы полезно построить модель для предсказывания предлагаемой зарплаты, ведь это достаточно важный фактор для поиска работы.
6. Большинство вакансий, где требуется полный рабочий день и полная занятость.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = '''(SELECT e.name, COUNT(*) cnt_vacancies --первое место по количеству вакансий.
               FROM vacancies v JOIN employers e ON v.employer_id=e.id
               GROUP BY e.name
               ORDER BY cnt_vacancies DESC
               LIMIT 1)
               UNION --пятое место по количеству вакансий.
               (SELECT e.name, COUNT(*) cnt_vacancies
               FROM vacancies v JOIN employers e ON v.employer_id=e.id
               GROUP BY e.name
               ORDER BY cnt_vacancies DESC
               OFFSET 4
               LIMIT 1)
               '''

In [23]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  cnt_vacancies
0  Газпром нефть            331
1         Яндекс           1933

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2 = '''SELECT a.name, COUNT(e.area) cnt_employers, COUNT(v.id) cnt_vacancies
               FROM areas a LEFT JOIN vacancies v ON a.id=v.area_id LEFT JOIN employers e ON e.area=a.id
               GROUP BY a.name 
               HAVING COUNT(v.id) = 0 --смотрим на те регионы, где нет вакансий.
               ORDER BY 2 DESC
               '''

In [25]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  cnt_employers  cnt_vacancies
0                          Россия            410              0
1                       Казахстан            207              0
2              Московская область             75              0
3              Краснодарский край             19              0
4              Ростовская область             18              0
..                            ...            ...            ...
588                      Струнино              0              0
589    Славгород (Алтайский край)              0              0
590                     Сорочинск              0              0
591  Ленинский (Тульская область)              0              0
592                         Тавда              0              0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_5_3='''SELECT e.name, COUNT(DISTINCT a.name) cnt_areas
             FROM vacancies v LEFT JOIN employers e ON e.id=v.employer_id LEFT JOIN areas a ON a.id=v.area_id 
             GROUP BY e.name
             ORDER BY 2 DESC
            '''

In [27]:
# результат запроса
df=pd.read_sql_query(query_5_3, connection)
df.head(15)

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  cnt_areas
0                       Яндекс        181
1                   Ростелеком        152
2                   Спецремонт        116
3       Поляков Денис Иванович         88
4                    ООО ЕФИН          71
5                   Совкомбанк         63
6                          МТС         55
7   ЭФКО, Управляющая компания         49
8                 Почта России         48
9                         КРОН         48
10                 ИК СИБИНТЕК         46
11                       MCORE         46
12             Сбербанк-Сервис         45
13                       ANCOR         44
14                    Тинькофф         43

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_5_4='''SELECT COUNT(*)
             FROM employers e LEFT JOIN employers_industries ei ON e.id=ei.employer_id LEFT JOIN industries i ON i.id=ei.industry_id
             WHERE i.id IS NULL
             '''

In [29]:
# результат запроса
df=pd.read_sql_query(query_5_4, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
query_5_5='''SELECT e.name
             FROM employers e LEFT JOIN employers_industries ei ON e.id=ei.employer_id LEFT JOIN industries i ON i.id=ei.industry_id
             GROUP BY e.name
             HAVING COUNT(i.id) = 4
             ORDER BY 1
             OFFSET 2
             LIMIT 1
             '''

In [31]:
# результат запроса
df=pd.read_sql_query(query_5_5,connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6='''SELECT COUNT(*)
             FROM employers e LEFT JOIN employers_industries ei ON e.id=ei.employer_id LEFT JOIN industries i ON i.id=ei.industry_id
             WHERE i.name = 'Разработка программного обеспечения'
             '''

In [33]:
# результат запроса
df=pd.read_sql_query(query_5_6, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup
#Парсим сайт с городами-миллионниками России
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
table = page.find('table', class_='standard') #Считываем таблицу гродов-миллионников
cities = [i.text for i in table.find_all('a')]
i = 0 
while i < len(cities): #Чистим получившийся список
    if not cities[i][0].isupper():
        del cities[i]
    else:
        i += 1
print(cities)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград']


In [35]:
# текст запроса
query_5_7 = '''SELECT a.name, COUNT(*) cnt
               FROM vacancies v LEFT JOIN areas a ON a.id=v.area_id LEFT JOIN employers e ON e.id=v.employer_id
               WHERE e.name='Яндекс' AND a.name IN '''+str(tuple(cities))+'''-- Смотрим только на города-миллионники.
               GROUP BY a.name
               UNION --Объединяем с total(суммарное количество вакансий).
               SELECT 'total', SUM(B.cnt)
               FROM
               (SELECT a.name, COUNT(*) cnt
               FROM vacancies v LEFT JOIN areas a ON a.id=v.area_id LEFT JOIN employers e ON e.id=v.employer_id
               WHERE e.name='Яндекс' AND a.name IN '''+str(tuple(cities))+'''
               GROUP BY a.name) AS B
               ORDER BY 2
               '''
               

In [36]:
# результат запроса
df=pd.read_sql_query(query_5_7, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7               Уфа   26.0
8            Самара   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            total  485.0

### Выводы по анализу работодателей:
1. Больше всего вакансий у крупных компаний, связанных с IT, банковским делом, или государственных компаний. Такая же тенденция с распространением компаний по регионам(у таких компаний есть вакансии в большом количестве регионов). 
2. Яндекс является ведущим работодателем в России как по количеству вакансий, так и по количеству регионов, где предлагают работу в данной компании. У него всего 1933 вакансий и 485 из них в городах-миллионниках.



***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
# текст запроса
query_6_1='''SELECT COUNT(*)
             FROM vacancies
             WHERE LOWER(name) LIKE '%data%' or LOWER(name) LIKE '%данн%'
             '''

In [38]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
# текст запроса
query_6_2='''SELECT COUnT(*)
             FROM vacancies
             WHERE (lower(name) LIKE '%data scientist%'
             OR lower(name) LIKE '%data science%' 
             OR lower(name) LIKE '%исследователь данных%' 
             OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
             OR lower(name) LIKE '%machine learning%'
             OR lower(name) LIKE '%машинн%обучен%') AND
             (lower(name) LIKE '%junior%' OR
             experience = 'Нет опыта' OR
             employment = 'Стажировка')
             '''

In [40]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
# текст запроса
query_6_3='''SELECT COUNT(*)
             FROM vacancies
             WHERE (lower(name) LIKE '%data scientist%'
             OR lower(name) LIKE '%data science%' 
             OR lower(name) LIKE '%исследователь данных%' 
             OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
             OR lower(name) LIKE '%machine learning%'
             OR lower(name) LIKE '%машинн%обучен%') AND
             (key_skills LIKE '%SQL%' OR
             key_skills LIKE '%postgres%')
             '''

In [42]:
# результат запроса
df=pd.read_sql_query(query_6_3, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
# текст запроса
query_6_4='''SELECT COUNT(*)
             FROM vacancies
             WHERE (lower(name) LIKE '%data scientist%'
             OR lower(name) LIKE '%data science%' 
             OR lower(name) LIKE '%исследователь данных%' 
             OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
             OR lower(name) LIKE '%machine learning%'
             OR lower(name) LIKE '%машинн%обучен%') AND
             key_skills LIKE '%Python%'
             '''

In [44]:
# результат запроса
df=pd.read_sql_query(query_6_4, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
# текст запроса
query_6_5='''SELECT
             ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), ''))+1), 2)
             --смотрим на количество разделителей между ключевыми навыками(в данных разделитель это "\t")
             from vacancies 
             WHERE (lower(name) LIKE '%data scientist%' OR 
             lower(name) LIKE '%data science%' OR
             lower(name) LIKE '%исследователь данных%' OR
             (name LIKE '%ML%' AND name NOT LIKE '%HTML%') OR
             lower(name) LIKE '%machine learning%' OR
             lower(name) LIKE '%машинн%обучен%')
             AND (key_skills IS NOT NULL)
             '''

In [46]:
# результат запроса
df=pd.read_sql_query(query_6_5, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
# текст запроса
# текст запроса
query_6_6='''SELECT ROUND(AVG((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from))/2))
             from vacancies 
             WHERE (lower(name) LIKE '%data scientist%' OR 
             lower(name) LIKE '%data science%' OR
             lower(name) LIKE '%исследователь данных%' OR
             (name LIKE '%ML%' AND name NOT LIKE '%HTML%') OR
             lower(name) LIKE '%machine learning%' OR
             lower(name) LIKE '%машинн%обучен%') AND
             experience='От 3 до 6 лет' AND
             (salary_from IS NOT NULL OR salary_to IS NOT NULL)
             '''

In [48]:
# результат запроса
df=pd.read_sql_query(query_6_6, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


round
0  243115.0

7. Проверим общее количество вакансий, связанных с data science

In [49]:
# текст запроса
query_6_7='''SELECT COUNT(*)
             FROM vacancies
             WHERE (lower(name) LIKE '%data scientist%'
             OR lower(name) LIKE '%data science%' 
             OR lower(name) LIKE '%исследователь данных%' 
             OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
             OR lower(name) LIKE '%machine learning%'
             OR lower(name) LIKE '%машинн%обучен%')
             '''

In [50]:
df=pd.read_sql_query(query_6_7, connection)
df.head()

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    480

8. Проверим среднее количество ключевых навыков для всех вакансий(чтобы сравнить со средним количество ключевых навыков для вакансий, связанных с data science)

In [51]:
# текст запроса
query_6_8='''SELECT
             ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), ''))+1), 2)
             from vacancies 
             WHERE key_skills IS NOT NULL
             '''

In [52]:
df=pd.read_sql_query(query_6_8, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


round
0   6.37

### Выводы по предметному анализу:
1. Примерно 3.6% всех вакансий составляют вакансии, связанные с даннымми.
2. Всего количество вакансий, связанных с data science - 480. Из них 51 вакансия для начинающих специалистов, что составляет примерно 10%.
3. Исходя из предыдущих наблюдений в среднем вакансий для начинающих специалистов больше в процентном соотношении от всех вакансий, что может говорить о том, что в сфере data science очень ценится опыт работника, и найти работу неопытному специалисту будет не так легко.
4. В большинстве вакансий, связанных с data science, требуется знание python, что логично, т.к. это один из самых(если не самый) популярных языков для работы с данными. 
5. Почти в половине вакансий требуется знание SQL и postgres, что говорит о том, что это очень важные инструменты для работников в сфере data science
 


***

# Общий вывод по проекту

1. Посмотрим какие самые популярные сферы деятельности, указанные в вакансиях

In [53]:
# текст запроса
query_7_1='''SELECT i.name, COUNT(*)
             FROM vacancies v LEFT JOIN employers e ON v.employer_id=e.id 
             LEFT JOIN employers_industries ei ON e.id=ei.employer_id 
             LEFT JOIN industries i ON i.id=ei.industry_id
             GROUP BY i.name
             ORDER BY 2 DESC
             '''

In [54]:
df=pd.read_sql_query(query_7_1, connection)
df.head(10)

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0                Разработка программного обеспечения  12499
1  Системная интеграция,  автоматизации технологи...  11034
2                                               None  10084
3  Интернет-компания (поисковики, платежные систе...   6413
4                                               Банк   2742
5                                    Мобильная связь   1585
6                                 Кадровые агентства   1143
7                              Консалтинговые услуги   1137
8                                 Интернет-провайдер   1095
9                       Розничная сеть (продуктовая)   1034

In [55]:
top_industries=list(df['name'].head(11)) #Сформируем список самых популярных сфер деятельности
print(top_industries)
del top_industries[2] #Удалим None
top_industries = tuple(top_industries)
print(top_industries)

['Разработка программного обеспечения', 'Системная интеграция,  автоматизации технологических и бизнес-процессов предприятия, ИТ-консалтинг', None, 'Интернет-компания (поисковики, платежные системы, соц.сети, информационно-познавательные и развлекательные ресурсы, продвижение сайтов и прочее)', 'Банк', 'Мобильная связь', 'Кадровые агентства', 'Консалтинговые услуги', 'Интернет-провайдер', 'Розничная сеть (продуктовая)', 'Фиксированная связь']
('Разработка программного обеспечения', 'Системная интеграция,  автоматизации технологических и бизнес-процессов предприятия, ИТ-консалтинг', 'Интернет-компания (поисковики, платежные системы, соц.сети, информационно-познавательные и развлекательные ресурсы, продвижение сайтов и прочее)', 'Банк', 'Мобильная связь', 'Кадровые агентства', 'Консалтинговые услуги', 'Интернет-провайдер', 'Розничная сеть (продуктовая)', 'Фиксированная связь')


2. Посмотрим среднее количество ключевых навыков, указанное в вакансиях в самых популярных сферах деятельности

In [56]:
# текст запроса
query_7_2=f'''SELECT i.name,
             ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), ''))+1), 2) avg_key_skills
             FROM vacancies v LEFT JOIN employers e ON v.employer_id=e.id 
             LEFT JOIN employers_industries ei ON e.id=ei.employer_id 
             LEFT JOIN industries i ON i.id=ei.industry_id
             WHERE i.name IN {top_industries}
             GROUP BY i.name
             ORDER BY 2 DESC
             '''

In [57]:
df=pd.read_sql_query(query_7_2, connection)
df.head(10)

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  avg_key_skills
0                              Консалтинговые услуги            7.35
1  Интернет-компания (поисковики, платежные систе...            6.84
2                                Фиксированная связь            6.71
3                                 Интернет-провайдер            6.70
4                                 Кадровые агентства            6.64
5  Системная интеграция,  автоматизации технологи...            6.63
6                                    Мобильная связь            6.51
7                Разработка программного обеспечения            6.46
8                       Розничная сеть (продуктовая)            6.05
9                                               Банк            6.03

3. Узнаем среднюю зарплату в вакансиях в самых популярных сфер деятельности в зависимости от опыта работы.

In [58]:
# текст запроса
query_7_3=f'''SELECT i.name, v.experience,
             ROUND(AVG((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from))/2)) AVG_salary
             FROM vacancies v LEFT JOIN employers e ON v.employer_id=e.id 
             LEFT JOIN employers_industries ei ON e.id=ei.employer_id 
             LEFT JOIN industries i ON i.id=ei.industry_id
             WHERE i.name IN {top_industries} AND
             (salary_from IS NOT NULL OR salary_to IS NOT NULL)
             GROUP BY i.name, v.experience
             ORDER BY 3 DESC
             '''

In [59]:
df_7_3=pd.read_sql_query(query_7_3, connection)
df_7_3

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name          experience  \
0                                  Кадровые агентства         Более 6 лет   
1                                  Интернет-провайдер         Более 6 лет   
2   Интернет-компания (поисковики, платежные систе...         Более 6 лет   
3                 Разработка программного обеспечения         Более 6 лет   
4                               Консалтинговые услуги         Более 6 лет   
5   Системная интеграция,  автоматизации технологи...         Более 6 лет   
6                                  Кадровые агентства       От 3 до 6 лет   
7                 Разработка программного обеспечения       От 3 до 6 лет   
8   Интернет-компания (поисковики, платежные систе...       От 3 до 6 лет   
9   Системная интеграция,  автоматизации технологи...       От 3 до 6 лет   
10                                               Банк         Более 6 лет   
11                                               Банк       От 3 до 6 лет   
12                              Консалтинговые услуги       От 3 до 6 лет   
13                                 Интернет-провайдер       От 3 до 6 лет   
14                                Фиксированная связь         Более 6 лет   
15                                 Кадровые агентства  От 1 года до 3 лет   
16                Разработка программного обеспечения  От 1 года до 3 лет   
17                                    Мобильная связь       От 3 до 6 лет   
18  Системная интеграция,  автоматизации технологи...  От 1 года до 3 лет   
19                       Розничная сеть (продуктовая)         Более 6 лет   
20                       Розничная сеть (продуктовая)       От 3 до 6 лет   
21                                    Мобильная связь         Более 6 лет   
22                              Консалтинговые услуги  От 1 года до 3 лет   
23                                Фиксированная связь       От 3 до 6 лет   
24  Интернет-компания (поисковики, платежные систе...  От 1 года до 3 лет   
25                                 Кадровые агентства           Нет опыта   
26                                               Банк  От 1 года до 3 лет   
27                       Розничная сеть (продуктовая)  От 1 года до 3 лет   
28                                    Мобильная связь  От 1 года до 3 лет   
29                                 Интернет-провайдер  От 1 года до 3 лет   
30                                Фиксированная связь  От 1 года до 3 лет   
31                              Консалтинговые услуги           Нет опыта   
32                                               Банк           Нет опыта   
33                       Розничная сеть (продуктовая)           Нет опыта   
34                Разработка программного обеспечения           Нет опыта   
35  Системная интеграция,  автоматизации технологи...           Нет опыта   
36                                 Интернет-провайдер           Нет опыта   
37                                    Мобильная связь           Нет опыта   
38                                Фиксированная связь           Нет опыта   
39  Интернет-компания (поисковики, платежные систе...           Нет опыта   

    avg_salary  
0     302154.0  
1     300000.0  
2     268118.0  
3     220904.0  
4     216593.0  
5     195760.0  
6     193969.0  
7     185022.0  
8     171830.0  
9     160288.0  
10    156886.0  
11    153075.0  
12    141245.0  
13    137784.0  
14    120000.0  
15    110660.0  
16     94860.0  
17     91238.0  
18     87098.0  
19     86940.0  
20     85972.0  
21     85250.0  
22     82402.0  
23     80729.0  
24     80607.0  
25     77335.0  
26     75438.0  
27     67351.0  
28     55943.0  
29     55846.0  
30     52502.0  
31     48657.0  
32     48417.0  
33     48130.0  
34     46070.0  
35     45016.0  
36     39484.0  
37     38949.0  
38     38510.0  
39     27626.0

4. Сделаем то же самое для вакансий, связанных с data science

In [60]:
# текст запроса
query_7_4=f'''SELECT experience exp, ROUND(AVG((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from))/2)) avg_salary
             from vacancies
             WHERE (lower(name) LIKE '%data scientist%' OR 
             lower(name) LIKE '%data science%' OR
             lower(name) LIKE '%исследователь данных%' OR
             (name LIKE '%ML%' AND name NOT LIKE '%HTML%') OR
             lower(name) LIKE '%machine learning%' OR
             lower(name) LIKE '%машинн%обучен%') AND
             (salary_from IS NOT NULL OR salary_to IS NOT NULL)
             GROUP BY experience
             ORDER BY 2 DESC
             '''

In [61]:
df_7_4=pd.read_sql_query(query_7_4, connection)
df_7_4

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


exp  avg_salary
0       От 3 до 6 лет    243115.0
1  От 1 года до 3 лет    139675.0
2           Нет опыта     74643.0

5. Посмотрим на среднюю зарплату в городах-миллионниках и население в них

In [62]:
# текст запроса
query_7_5=f'''SELECT a.name, ROUND(AVG((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from))/2)) AVG_salary 
             FROM vacancies v JOIN areas a ON v.area_id=a.id
             WHERE a.name IN {str(tuple(cities))}
             GROUP BY a.name
             HAVING ROUND(AVG((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from))/2)) IS NOT NULL
             ORDER BY 1 --Отсортируем по названию городов, чтобы далее легче было соотнести город и его население.
             '''

In [63]:
df=pd.read_sql_query(query_7_5, connection)
df

C:\Users\mrcho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  avg_salary
0         Волгоград     68689.0
1           Воронеж     84958.0
2      Екатеринбург     91896.0
3            Казань     92207.0
4         Краснодар     92328.0
5        Красноярск     75156.0
6            Москва    134140.0
7   Нижний Новгород     96059.0
8       Новосибирск    107777.0
9              Омск     86393.0
10            Пермь     95985.0
11   Ростов-на-Дону     94065.0
12           Самара     80223.0
13  Санкт-Петербург    104410.0
14              Уфа     74733.0
15        Челябинск     77463.0

In [64]:
#Парсим сайт с городами-миллионниками России
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
table = page.find('table', class_='standard') #Считываем таблицу городов-миллионников
table_elems = [i.text for i in table.find_all('td')] #
cities_with_popul={} #Создадим словарь для записи городов-миллионников и их населения
for i,elems in enumerate(table_elems):
    if elems in cities:
        cities_with_popul[elems] = table_elems[i+1]
    elif elems[:-3] in cities: #Отдельный случай для записей типа Воронеж[а]
        cities_with_popul[elems[:-3]] = table_elems[i+1]
#Отсортируем наш словарь по ключам, чтобы соотнести города в нашей таблице и их население
cities_with_popul = dict(sorted(cities_with_popul.items()))  
print(cities_with_popul)

{'Волгоград': '1028', 'Воронеж': '1058', 'Екатеринбург': '1544', 'Казань': '1309', 'Краснодар': '1099', 'Красноярск': '1188', 'Москва': '13 010', 'Нижний Новгород': '1226', 'Новосибирск': '1634', 'Омск': '1126', 'Пермь': '1034', 'Ростов-на-Дону': '1142', 'Самара': '1173', 'Санкт-Петербург': '5602', 'Уфа': '1145', 'Челябинск': '1190'}


In [65]:
df['population'] = cities_with_popul.values() #Создаём отдельный столбец с населением(в тыс. человек)
df

name  avg_salary population
0         Волгоград     68689.0       1028
1           Воронеж     84958.0       1058
2      Екатеринбург     91896.0       1544
3            Казань     92207.0       1309
4         Краснодар     92328.0       1099
5        Красноярск     75156.0       1188
6            Москва    134140.0     13 010
7   Нижний Новгород     96059.0       1226
8       Новосибирск    107777.0       1634
9              Омск     86393.0       1126
10            Пермь     95985.0       1034
11   Ростов-на-Дону     94065.0       1142
12           Самара     80223.0       1173
13  Санкт-Петербург    104410.0       5602
14              Уфа     74733.0       1145
15        Челябинск     77463.0       1190

In [66]:
#Отсортируем по зарплате и населению
df = df.sort_values( ['avg_salary','population'], ascending=[False,False], ignore_index=True)
df

name  avg_salary population
0            Москва    134140.0     13 010
1       Новосибирск    107777.0       1634
2   Санкт-Петербург    104410.0       5602
3   Нижний Новгород     96059.0       1226
4             Пермь     95985.0       1034
5    Ростов-на-Дону     94065.0       1142
6         Краснодар     92328.0       1099
7            Казань     92207.0       1309
8      Екатеринбург     91896.0       1544
9              Омск     86393.0       1126
10          Воронеж     84958.0       1058
11           Самара     80223.0       1173
12        Челябинск     77463.0       1190
13       Красноярск     75156.0       1188
14              Уфа     74733.0       1145
15        Волгоград     68689.0       1028

In [67]:
connection.close()

### Дополнительные и общие выводы:
1. В среднем количество ключевых навыков в вакансиях в самых популярных сферах деятельности примерно варьируется от 6 до 8.
2. Можно заметить, что в среднем зарплаты data scientist'ов больше зарплаты специалистов в самых популярных сферах деятельности.
3. Население не всегда напрямую влияет на зарплату. Это стоит учитывать при создании модели, предсказывающей зарплату вакансии. 
4. В дальнейшем можно будет детальнее изучить зависимости различных признаков. Можно, например, выделить такой новый признак, как площадь города. В данных представлены вакансии, где не указана сфера деятельности, поэтому можно попытаться придумать алгоритм, с помощью которого можно будет заполнить этот признак до конца и желательно нужно будет заполнить пробелы в в таком важном признаке, как зарплата, т.к. это очень важный признак для модели, которая рекомендует вакансии. 